include/rdma/fi_profile.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2023 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_PROFILE_H
+#define FI_PROFILE_H
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <rdma/fi_domain.h>
+
+/*
+ * pre-defined profiling variables
+ */
+enum {
+	FI_VAR_UNEXP_MSG_CNT,      // datatype: FI_UINT64
+	FI_VAR_UNEXP_MSG_QUEUE,    // datatype: FI_TYPE_CQ_ERR_ENTRY
+};
+
+/*
+ * pre-defined events which can change the profiling variables
+ */
+enum {
+	FI_EVENT_UNEXP_MSG_RECVD,    // var = FI_VAR_UNEXP_MSG_CNT
+	FI_EVENT_UNEXP_MSG_MATCHED,  // var = FI_VAR_UNEXP_MSG_CNT
+};
+
+struct fi_profile_desc {
+	uint32_t id;
+	union {
+		enum fi_datatype basic;
+		enum fi_type comp;
+	} type;
+	uint64_t flags;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+
+#define FI_VAR_TYPE_COMP  0x1   //  flags bit0,  0: basic, 1: comp
+
+struct fid_profile;
+
+struct fi_profile_ops {
+	size_t size;
+	void (*reset)(struct fid_profile *prof_fid,  uint64_t flags);
+	ssize_t (*query_vars)(struct fid_profile *prof_fid,
+	                      struct fi_profile_desc *varlist, size_t *count);
+	ssize_t (*query_events)(struct fid_profile *prof_fid,
+	                       struct fi_profile_desc *eventlist, size_t *count);
+	ssize_t (*read_var)(struct fid_profile *prof_fid, uint32_t var_id,
+	                    void *data, size_t *size);
+	int (*reg_callback)(struct fid_profile *prof_fid, uint32_t event,
+	        int (*callback)(struct fid_profile *prof_fid, uint32_t event, 
+	                void *param, size_t size, uint32_t type, void *context),
```

Comment:
I'm unsure about the type parameter here.  Now that type could be fi_datatype or fi_type, it's unclear.  I think the user will just need to know/get this from the fi_profile_desc that's associated with the event id.

Hmm... maybe instead of passing the u32 event to the callback, we should just pass struct fi_profile_desc *?  That includes the event, plus type information, name, etc., which avoids the caller needing to look it up via the context parameter.

Reply:
We can use fi_profile_desc instead event_Id in the callback,  here the size and datatype in the fi_profile_desc will be for the data (param) of callback for that event.  We can pass fi_profile_desc for variable in read_var(),  this can reduce the looking up as well.

Reply:
The provider owns the internal allocation of fi_profile_desc and the event id.  It's trivial for the provider to lookup the fi_profile_desc given an id.  I would leave read_var() as-is.

Having the provider hand fi_profile_desc to the user via the callback makes sense.  But the reverse isn't true (user passing an fi_profile_desc to the provider).